Shift3/boilerplate-client-react
# Add logout modal

## Changes
1. Added LogoutModal component.
2. Added custom hook to return modal state management and callbacks.
3. Render LogoutModal in the TopNav when the "Sign Out" button is clicked in the profile dropdown menu.

## Purpose
When a user clicks the sign out button, the application should open a modal to confirm that the user wants to sign out and end the current session.

## Approach
- The `react-bootstrap` Modal component is used since it already implements responsiveness and accessibility features.
- Management of the modals state and callback functions are moved into a custom hook to make reusing the modal easier without having to redefine state and callback functions (e.g. it will be re-used in the SideNav component as well in the future). 
- The `react-bootstrap` Modal component uses React Portals under the hood, so even though the `LogoutModal` is being rendered inside the `TopNav`, the `LogoutModal` will actually be mounted to a separate root level subtree outside the DOM hierarchy of the `TopNav`. e.g. it will actually be mounted as follows
```jsx
<div id="app">
    ...
    <TopNav/>
    ...
</div>

<div id="modal-root">
    <LogoutModal />
</div>
```

## Pre-Testing TODOs
1. Have the `boilerplate-server-node`  running and the seeded account activated.

## Testing

1. Pull in the changes to your local copy of this branch.
2. Run `yarn start` to start the app.
3. Navigate to `localhost:4200/auth/login` and login with the credentials of the seeded user.
4. Once logged in, confirm that clicking the "Sign Out" button in the dropdown menu opens a modal as shown in the screenshot below.
5. Confirm that clicking the "CANCEL" modal button in the modal closes the modal without logging the user out.
6. Confirm that clicking anywhere outside of the modal closes the modal without logging the user out.
7. Confirm that clicking the "LOGOUT" modal button closes the modal and logs the user out.
8. Confirm that you are able to use the `TAB` key to cycle through the modal buttons.
9. Confirm that when the modal is open, using the `TAB` key keep only cycles through the modal buttons and the url bar.

## Screenshots
![Screen Shot 2021-07-19 at 10 19 45 AM](https://user-images.githubusercontent.com/40707951/126201913-7812e583-16c6-4606-b0bb-8f0b090a029e.png)

### Closes #117
### Closes #120
